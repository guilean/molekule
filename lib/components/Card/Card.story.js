import React from 'react';
import { storiesOf } from '@storybook/react';
import Card from './Card';

storiesOf('Card', module).add('Basic', () => (
  <Card>
    <Card.Header>Header</Card.Header>
    <Card.Body>This is the body of the card</Card.Body>
    <Card.Footer>Footer</Card.Footer>
  </Card>
));
