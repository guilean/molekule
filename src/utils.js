import get from 'lodash/get';
import kebabCase from 'lodash/kebabCase';
import styled from 'styled-components';

export const getFromTheme = (theme, lookup, fallback) => get(theme, lookup, fallback);

export const getComponentVariant = (theme, key, variant) => {
  const config = getFromTheme(theme, `components.${key}.variants.${variant}`, theme.variants[variant]);
  if (!config) throw new Error(`Refractal: "${variant}" variant not found in theme...`);
  return config;
};

export const getComponentStyle = key => ({ theme }) => getFromTheme(theme, `components.${key}.style`, {});

export const getComponentClassName = name => ({ theme: { classPrefix }, variant }) =>
  `${classPrefix}-${name} ${variant ? `${classPrefix}-${name}-${variant}` : ''}`.trim();

export const createComponent = ({ name, tag = 'div', as }) => {
  const component = as ? styled(as) : styled[tag];

  return component.attrs({
    className: getComponentClassName(kebabCase(name)),
  })`
    ${getComponentStyle(name)}
    ${({ styles = {} }) => styles[name] || {}}
  `;
};
