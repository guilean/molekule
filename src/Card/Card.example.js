import React from 'react';
import Card from './Card';
import LiveEdit from '../live-edit';

export default {
  group: 'Card',
  render: () => (
    <React.Fragment>
      <p>
        Cards provide a flexible way to encapsulate content with multiple variants and options.
      </p>

      <LiveEdit
        code={`<Card>
  <Card.Header>Header</Card.Header>
  <Card.Body>A card is a flexible and extensible content container. It includes options for headers and footers, a wide variety of content, contextual background colors, and powerful display options.</Card.Body>
  <Card.Footer>Footer</Card.Footer>
</Card>`}
        scope={{ Card }}
        transformCode={code => `<div style={{ maxWidth: '300px' }}>${code}</div>`}
      />
    </React.Fragment>
  ),
};
