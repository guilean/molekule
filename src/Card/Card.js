import React from 'react';
import PropTypes from 'prop-types';
import { css } from 'styled-components';
import Box from '../Box';
import { themeGet, createComponent } from '../utils';

const StyledCard = createComponent({
  name: 'Card',
  as: Box,
  style: ({ shadow, theme }) => css`
    background: white;
    box-shadow: ${shadow ? themeGet('shadow.soft', 'none') : 'none'};
    border-radius: ${themeGet('radius', 0)}px;
    overflow: hidden;
    border: 1px solid ${theme.colors.greyLight};
  `,
});

/** Cards provide a flexible way to encapsulate content with multiple variants and options. */
const Card = React.forwardRef((props, ref) => <StyledCard ref={ref} {...props} />);

Card.propTypes = {
  shadow: PropTypes.bool,
};

Card.defaultProps = {
  shadow: false,
};

Card.Footer = createComponent({
  name: 'CardFooter',
  as: Box,
  style: css`
    padding: 16px;
  `,
});

Card.Body = createComponent({
  name: 'CardBody',
  as: Box,
  style: () => css`
    padding: 16px;

    & + ${Card.Footer} {
      padding-top: 0px;
    }
  `,
});

Card.Header = createComponent({
  name: 'CardHeader',
  as: Box,
  style: css`
    padding: 16px;
    font-weight: 700;

    & + ${Card.Body} {
      padding-top: 0px;
    }
  `,
});

export default Card;
