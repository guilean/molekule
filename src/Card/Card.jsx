import React from 'react';
import PropTypes from 'prop-types';
import { space, borderRadius, themeGet } from 'styled-system';
import { createComponent } from '../utils';

const StyledCard = createComponent({
  name: 'Card',
}).extend`
  background: white;
  box-shadow: ${p =>
    p.shadow ? themeGet('shadow', '0 4px 10px rgba(0, 29, 54, 0.04), 0 -2px 6px rgba(0, 29, 54, 0.025);') : 'none'};

  ${borderRadius};
  ${space};
`;

const Card = props => <StyledCard {...props} />;

Card.propTypes = {
  shadow: PropTypes.bool,
};

Card.defaultProps = {
  shadow: true,
};

Card.Header = createComponent({
  name: 'CardHeader',
}).extend`
  padding: 1rem;
  border-bottom: 1px solid ${themeGet('colors.grayLight')};

  ${space};
`;

Card.Body = createComponent({
  name: 'CardBody',
}).extend`
  padding: 1rem;

  ${space};
`;

Card.Footer = createComponent({
  name: 'CardFooter',
}).extend`
  padding: 1rem;
  border-top: 1px solid ${themeGet('colors.grayLight')};

  ${space};
`;

export default Card;
