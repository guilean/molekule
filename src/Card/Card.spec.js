import React from 'react';
import { mountWithTheme } from '../../test/utils';
import Card from './Card';

test('Card', () => {
  const component = mountWithTheme(
    <Card>
      <Card.Header>Header</Card.Header>
      <Card.Body>Body</Card.Body>
      <Card.Footer>Footer</Card.Footer>
    </Card>
  );

  expect(component).toMatchSnapshot();
});
