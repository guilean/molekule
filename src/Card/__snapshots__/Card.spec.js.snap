// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card 1`] = `
<DocumentFragment>
  .c0 {
  background: white;
  box-shadow: none;
  border-radius: 8px;
  overflow: hidden;
  border: 1px solid #E8E9EA;
}

.c5 {
  padding: 16px;
}

.c3 {
  padding: 16px;
}

.c3 + .c4 {
  padding-top: 0px;
}

.c1 {
  padding: 16px;
  font-weight: 700;
}

.c1 + .c2 {
  padding-top: 0px;
}

<div
    class="re-box re-card c0"
  >
    <div
      class="re-box re-card-header c1"
    >
      Header
    </div>
    <div
      class="c2 re-box re-card-body c3"
    >
      Body
    </div>
    <div
      class="c4 re-box re-card-footer c5"
    >
      Footer
    </div>
  </div>
</DocumentFragment>
`;
